From 3451d2fe2e5e790b72eb9fbab7fb40836cfe0b4d Mon Sep 17 00:00:00 2001
From: Emily Shepherd <emily@redcoat.dev>
Date: Tue, 14 Feb 2023 14:53:16 +0000
Subject: [PATCH] Delete pidfile support

As chronyd will be running within a container in the kiOS eco-system,
there is no benefit in writing a pid file. In fact, it can cause more
trouble, as it requires having a section of the file system which is
writable.

For this project we can simply remove this functionality entirely.
---
 conf.c | 16 --------------
 conf.h |  1 -
 main.c | 67 ----------------------------------------------------------
 3 files changed, 84 deletions(-)

diff --git a/conf.c b/conf.c
index 8dd2743..9fc345e 100644
--- a/conf.c
+++ b/conf.c
@@ -211,10 +211,6 @@ static int ntp_dscp = 0;
 /* Path to Samba (ntp_signd) socket. */
 static char *ntp_signd_socket = NULL;
 
-/* Filename to use for storing pid of running chronyd, to prevent multiple
- * chronyds being started. */
-static char *pidfile = NULL;
-
 /* Rate limiting parameters */
 static int ntp_ratelimit_enabled = 0;
 static int ntp_ratelimit_interval = 3;
@@ -408,7 +404,6 @@ CNF_Initialise(int r, int client_only)
     cmd_port = ntp_port = 0;
   } else {
     bind_cmd_path = Strdup(DEFAULT_COMMAND_SOCKET);
-    pidfile = Strdup(DEFAULT_PID_FILE);
   }
 
   SCK_GetAnyLocalIPAddress(IPADDR_INET4, &bind_address4);
@@ -471,7 +466,6 @@ CNF_Finalise(void)
   Free(bind_cmd_iface);
   Free(bind_cmd_path);
   Free(ntp_signd_socket);
-  Free(pidfile);
   Free(rtc_device);
   Free(rtc_file);
   Free(user);
@@ -684,8 +678,6 @@ CNF_ParseLine(const char *filename, int number, char *line)
     parse_ntstrustedcerts(p);
   } else if (!strcasecmp(command, "peer")) {
     parse_source(p, command, 1);
-  } else if (!strcasecmp(command, "pidfile")) {
-    parse_string(p, &pidfile);
   } else if (!strcasecmp(command, "pool")) {
     parse_source(p, command, 1);
   } else if (!strcasecmp(command, "port")) {
@@ -2338,14 +2330,6 @@ CNF_GetNtpSigndSocket(void)
 
 /* ================================================== */
 
-char *
-CNF_GetPidFile(void)
-{
-  return pidfile;
-}
-
-/* ================================================== */
-
 REF_LeapMode
 CNF_GetLeapSecMode(void)
 {
diff --git a/conf.h b/conf.h
index 11fd11d..e6f3def 100644
--- a/conf.h
+++ b/conf.h
@@ -86,7 +86,6 @@ extern char *CNF_GetBindCommandInterface(void);
 extern char *CNF_GetBindCommandPath(void);
 extern int CNF_GetNtpDscp(void);
 extern char *CNF_GetNtpSigndSocket(void);
-extern char *CNF_GetPidFile(void);
 extern REF_LeapMode CNF_GetLeapSecMode(void);
 extern char *CNF_GetLeapSecTimezone(void);
 
diff --git a/main.c b/main.c
index c40b5e4..2817558 100644
--- a/main.c
+++ b/main.c
@@ -92,20 +92,6 @@ do_platform_checks(void)
 
 /* ================================================== */
 
-static void
-delete_pidfile(void)
-{
-  const char *pidfile = CNF_GetPidFile();
-
-  if (!pidfile)
-    return;
-
-  if (!UTI_RemoveFile(NULL, pidfile, NULL))
-    ;
-}
-
-/* ================================================== */
-
 void
 MAI_CleanupAndExit(void)
 {
@@ -142,8 +128,6 @@ MAI_CleanupAndExit(void)
   LCL_Finalise();
   PRV_Finalise();
 
-  delete_pidfile();
-  
   CNF_Finalise();
   HSH_Finalise();
   LOG_Finalise();
@@ -263,51 +247,6 @@ post_init_rtc_hook(void *anything)
 
 /* ================================================== */
 
-static void
-check_pidfile(void)
-{
-  const char *pidfile = CNF_GetPidFile();
-  FILE *in;
-  int pid, count;
-  
-  if (!pidfile)
-    return;
-
-  in = UTI_OpenFile(NULL, pidfile, NULL, 'r', 0);
-  if (!in)
-    return;
-
-  count = fscanf(in, "%d", &pid);
-  fclose(in);
-  
-  if (count != 1)
-    return;
-
-  if (getsid(pid) < 0)
-    return;
-
-  LOG_FATAL("Another chronyd may already be running (pid=%d), check %s",
-            pid, pidfile);
-}
-
-/* ================================================== */
-
-static void
-write_pidfile(void)
-{
-  const char *pidfile = CNF_GetPidFile();
-  FILE *out;
-
-  if (!pidfile)
-    return;
-
-  out = UTI_OpenFile(NULL, pidfile, NULL, 'W', 0644);
-  fprintf(out, "%d\n", (int)getpid());
-  fclose(out);
-}
-
-/* ================================================== */
-
 #define DEV_NULL "/dev/null"
 
 static void
@@ -588,9 +527,6 @@ int main
   if (print_config)
     return 0;
 
-  /* Check whether another chronyd may already be running */
-  check_pidfile();
-
   if (!user)
     user = CNF_GetUser();
 
@@ -601,9 +537,6 @@ int main
   /* Create directories for sockets, log files, and dump files */
   CNF_CreateDirs(pw->pw_uid, pw->pw_gid);
 
-  /* Write our pidfile to prevent other instances from running */
-  write_pidfile();
-
   PRV_Initialise();
   LCL_Initialise();
   SCH_Initialise();
-- 
2.39.1

